kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/AdminManager.java

```diff
@@ -212,7 +214,8 @@ public void shutdown() {
                                 future.complete(new DescribeConfigsResponse.Config(ApiError.NONE, dummyConfig));
                                 break;
                             default:
-                                throw new InvalidRequestException("Unsupported resource type: " + resource.type());
+                                return FutureUtil.failedFuture(
+                                        new InvalidRequestException("Unsupported resource type: " + resource.type()));
                         }
```

Comment:
I believe that throwing an Exception in a method that returns a CompletableFuture is a code smell and also it leads to bugs
because the caller, generally speaking, does not expect the method to throw anything.

in this case we are happy and we know the caller, but this is not always the case

so throwing such unchecked exception in methods that return a CompletableFuture is a source of hidden problems, that are very hard to track when they happen

I strongly believe that it is better to fix this.

@BewareMyPower

Reply:
> I believe that throwing an Exception in a method that returns a CompletableFuture is a code smell and also it leads to bugs

Yeah, here we should catch `InvalidRequestException` instead of `Exception`.

My concern is, whether the response future is completed exceptionally makes a difference in KoP. See `KafkaCommandDecoder#writeAndFlushResponseToClient`.

```java
            // case 2: responseFuture is completed exceptionally
            if (responseFuture.isCompletedExceptionally()) {
                responseFuture.exceptionally(e -> {
                    log.error("[{}] request {} completed exceptionally", channel, request.getHeader(), e);
                    channel.writeAndFlush(request.createErrorResponse(e));

                    requestStats.getStatsLogger()
                            .scopeLabel(KopServerStats.REQUEST_SCOPE,
                                    responseAndRequest.request.getHeader().apiKey().name)
                            .getOpStatsLogger(KopServerStats.REQUEST_QUEUED_LATENCY)
                            .registerFailedEvent(MathUtils.elapsedNanos(responseAndRequest.getCreatedTimestamp()),
                                    TimeUnit.NANOSECONDS);
                    return null;
                }); // send exception to client?
                continue;
            }
```

The original code completes the future normally while this PR will complete the future exceptionally if the resource type is unknown. It's more like a client side error if the resource type is unknown. Should we print the full stack and record it as failed events?

Reply:
I see your point.

it is better to do this in this point, instead of throwing the Exception and then returning the completedFuture:

```
return CompletableFuture.completedFuture(
                                new DescribeConfigsResponse.Config(ApiError.fromThrowable(new InvalidRequestException("Unsupported resource type: " + resource.type()), Collections.emptyList()));
```

this way it is clearer from my point of view, as we are not throwing exceptions from a method that is supposed to report errors as failed Futures.

does this make sense to you ?

Reply:
Yes